torch/_dynamo/variables/optimizer.py

```diff
@@ -60,6 +60,14 @@ def call_method(
                 ret_val = self.value._init_group(*py_args, **py_kwargs)
                 self.map_sources_and_install_guards(tx)
                 self.update_list_args(tx, args, kwargs, py_args, py_kwargs)
+                # this next line has the side effect of installing guards
```

Comment:
Can you explain this comment a little more, I don't see why this needs to be the case. 

We install guards on all the param groups previously, why does this order matter?

Reply:
See the OP. The example where this broke was
```
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_optim.py -k test_can_load_older_state_dict_ASGD_cpu_float32
```

The issue was that the internals of the optimizer had changed, so `___dict_keys_getitem(d, 1)` was now the learning rate rather than `params`, and we were installing first guards of the form
`___dict_keys_getitem(d, 1)[0] is <some other tensor>` before checking that `___dict_keys_getitem(d, 1)` was even a list

Reply:
We would have to construct the list during recursive_realize though right?

Reply:
So this is a bit of a corner case bug of https://github.com/pytorch/pytorch/pull/117625. In
```
                    GetItemSource(group_source, "params"), p_ind
```
we define the source of some parameters using the string `"params"`, while when tracing a dict, we do so by using the more verbose `___dict_keys_getitem(L['self'].param_groups[0], 5)` (this expression evaluates to `"params"`). Then, when calling `VariableBuilder` on `param_groups`, it looks at the two sources for these tensors and it thinks that they are different, so it eagerly places an unnecessary `make_dupe_guard`. This guard is put there before the other guards on the types are put in place, so we hit an error.

I'll add a PR into this stack that removes this guard entirely, and I will revert these changes then.

Reply:
Added a patch fixing these unnecessary guards (and pretty much reverting this PR) at https://github.com/pytorch/pytorch/pull/118535.